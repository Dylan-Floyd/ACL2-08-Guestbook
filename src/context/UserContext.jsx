import { createContext, useContext, useState } from 'react'

const UserContext = createContext({})

export const UserProvider = ({ children }) => {
  const [user, setUser] = useState({})

  return (
    <UserContext.Provider value={{ user, setUser }}>
      {children}
    </UserContext.Provider>
  )
}

export const useUser = () => {
  const context = useContext(UserContext)
  
  if (context === undefined) {
    throw new Error('useUser can only be called from a child component of UserProvider')
  }

  return context;
}