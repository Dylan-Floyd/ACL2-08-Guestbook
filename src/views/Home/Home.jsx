import { useState } from 'react'
import EntryForm from '../../components/EntryForm/EntryForm.jsx'
import EntryList from '../../components/EntryList/EntryList.jsx'
import { useUser } from '../../context/UserContext.jsx'

export default function Home() {
  const { user, setUser } = useUser()
  const [name, setName] = useState('')
  const [entry, setEntry] = useState('')
  const [entries, setEntries] = useState([])

  const isSignedIn = user?.name !== undefined

  const handleSubmit = (e) => {
    e.preventDefault();
    if(!isSignedIn) setUser({ name: name })
    setEntries((prevEntries) => [...prevEntries, { text: entry, name: name } ])
    setEntry('')
  }

  const handleSignout = (e) => {
    setUser({})
    setName('')
  }

  return (
    <div className="home">
      <EntryForm 
        {...{
          name,
          entry,
          isSignedIn,
          setName,
          setEntry,
          handleSubmit,
          handleSignout
        }}
      />
      <div>
        <EntryList entries={entries}/>
      </div>
    </div>
  )
}
