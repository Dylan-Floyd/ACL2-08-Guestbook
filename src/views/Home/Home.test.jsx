import { 
  screen,
  render,
  waitForElementToBeRemoved,
  within,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { UserProvider } from '../../context/UserContext.jsx';
import Home from './Home.jsx';

it('Renders the Home view and adds an entry', async () => {
  const { container } = render(
    <UserProvider>
      <Home />
    </UserProvider>
  )

  const nameInput = screen.getByLabelText(/guest name/i)
  const entryInput = screen.getByLabelText(/guest entry/i)
  const signButton = screen.getByLabelText('sign')

  userEvent.type(nameInput, 'test name')
  userEvent.type(entryInput, 'test entry')
  userEvent.click(signButton)

  expect(nameInput).not.toBeInTheDocument()
  expect(entryInput).toHaveValue('')

  const entry = await screen.findByText(/test entry/i)
  expect(entry).toBeInTheDocument()

  const newMessage = await screen.findByText(/thank you for signing/i)
  expect(newMessage).toBeInTheDocument()

  const signOutButton = await screen.findByLabelText('sign-out')
  expect(signOutButton).toBeInTheDocument()

  expect(container).toMatchSnapshot()
})