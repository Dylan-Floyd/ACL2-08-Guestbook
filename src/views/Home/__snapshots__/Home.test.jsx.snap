// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the Home view and adds an entry 1`] = `
<div>
  <div
    class="home"
  >
    <form
      class="entry-form"
    >
      <label>
        Thank you for signing 
        test name
        . Feel free to add more entries.
      </label>
      <label>
        Guest Entry
        <textarea
          rows="3"
        />
      </label>
      <button
        aria-label="sign"
        type="submit"
      >
        Sign
      </button>
      <button
        aria-label="sign-out"
      >
        Not 
        test name
        ?
      </button>
    </form>
    <div>
      <div>
        <ul>
          <div>
            test entry
            <br />
            <em>
              -
              test name
            </em>
          </div>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
